### DISCUSSION TITLE
How to use migration correctly?

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2432] [Q&A]
How to create my database collections? Should I do after changed collection field?

I wrote functions that create collection and called it in migration file. When I start an app, its creating my collections. 

So how to update collection after changed create collection function. Should I write code that update collection inside migration file?

**Comment by ganigeorgiev** [comment]
There are not enough details to understand what you mean and what you are trying to do.

If you haven't already, you can check the [Migrations docs](https://pocketbase.io/docs/migrations/) and enable the `Automigrate` so that any change from the UI will generate automatically a migration file for you.

> So how to update collection after changed create collection function. Should I write code that update collection inside migration file?

In general yes, but again, I'm not sure that I understand what you are trying todo. For collections update, you can check - https://pocketbase.io/docs/collection-methods/#update-existing-collection.

**Reply by Oggy4th** [reply]
I don't wanna use `Automigrate`.

I have functions like this one that create collection.

```
func CreateBranchCollection(txDao *daos.Dao, employeeCollectionId string) (*string, error) {
	collection := &models.Collection{
		Name:       "branches",
		Type:       models.CollectionTypeBase,
		ListRule:   types.Pointer("@request.auth.id != ''"),
		ViewRule:   types.Pointer("@request.auth.id != ''"),
		CreateRule: types.Pointer("@request.auth.id != ''"),
		UpdateRule: types.Pointer("@request.auth.id != ''"),
		DeleteRule: nil,

		Schema: schema.NewSchema(
			&schema.SchemaField{
				Name:     "name",
				Type:     schema.FieldTypeText,
				Required: true,
				Options: &schema.TextOptions{
					Max: types.Pointer(255),
				},
			},
			&schema.SchemaField{
				Name:     "status",
				Type:     schema.FieldTypeSelect,
				Required: false,
				Options: &schema.SelectOptions{
					MaxSelect: 1,
					Values:    []string{"active", "inactive"},
				},
			},
			&schema.SchemaField{
				Name:     "issuerId",
				Type:     schema.FieldTypeRelation,
				Required: true,
				Options: &schema.RelationOptions{
					CollectionId:  employeeCollectionId,
					CascadeDelete: false,
					MaxSelect:     types.Pointer(1),
					DisplayFields: []string{"id", "lastName", "firstName"},
				},
			},
		),
	}

	if err := txDao.SaveCollection(collection); err != nil {
		return nil, err
	}

	return &collection.Id, nil
}
```

And I called this functions inside migration file. When I start app its creating all my collections.
```
func init() {
	m.Register(func(db dbx.Builder) error {
		dao := daos.New(db)
		err := dao.RunInTransaction(func(txDao *daos.Dao) error {
			employeeCollectionId, err := employee.CreateEmployeeCollection(txDao)
			if err != nil {
				return err
			}

			branchCollectionId, err := branch.CreateBranchCollection(txDao, *employeeCollectionId)
			if err != nil {
				return err
			}

			return nil
		})

		if err != nil {
			return err
		}

		return nil
	}, func(db dbx.Builder) error {
		// add down queries...

		return nil
	})
}
```
I am testing it in server.

When I change status field how to apply this change to server?

**Reply by ganigeorgiev** [reply]
But what do you mean by _"When I change status field how to apply this change to server?"_
How do you change this field in the first place - from the UI or you want to change it manually?

If you don't want to use the `Automigrate` and want to change it manually, then you have to create a new migration that will update the previously inserted collections data using https://pocketbase.io/docs/collection-methods/#update-existing-collection.

**Reply by Oggy4th** [reply]
To change it manually using migration

**Reply by ganigeorgiev** [reply]
Here is an example:

```go
m.Register(func(db dbx.Builder) error {
	dao := daos.New(db)

	collection, err := dao.FindCollectionByNameOrId("branches")
	if err != nil {
	    return err
	}

	// this returns a pointer so we can directly modify the underlaying field data
	statusField := collection.Schema.GetFieldByName("status")
	statusField.Required = true

	// add a new field
	collection.Schema.AddField(&schema.SchemaField{
		Name: "exampleNewField",
		Type: schema.FieldTypeText,
	})

	// remove existing field
	issuerField := collection.Schema.GetFieldByName("issuerId")
	collection.Schema.RemoveField(issuerField.Id)

	return dao.SaveCollection(collection)
}, func(db dbx.Builder) error {
	// optionally, revert the above changes...

	return nil
})
```
